recommenders/models/sasrec/util.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import sys
+import copy
+import random
+import numpy as np
+import tensorflow as tf
+from collections import defaultdict
+from tqdm import tqdm
+import time
+
+
+def data_partition(fname):
+    usernum = 0
+    itemnum = 0
+    User = defaultdict(list)
+    user_train = {}
+    user_valid = {}
+    user_test = {}
+    # assume user/item index starting from 1
+    # f = open('data/%s.txt' % fname, 'r')
+    f = open(fname, "r")
+    for line in f:
+        u, i = line.rstrip().split(" ")
+        u = int(u)
+        i = int(i)
+        usernum = max(u, usernum)
+        itemnum = max(i, itemnum)
+        User[u].append(i)
+
+    for user in User:
+        nfeedback = len(User[user])
+        if nfeedback < 3:
+            user_train[user] = User[user]
+            user_valid[user] = []
+            user_test[user] = []
+        else:
+            user_train[user] = User[user][:-2]
+            user_valid[user] = []
+            user_valid[user].append(User[user][-2])
+            user_test[user] = []
+            user_test[user].append(User[user][-1])
+    return [user_train, user_valid, user_test, usernum, itemnum]
+
+
+def data_partition_with_time(fname, sep=" "):
+    usernum = 0
+    itemnum = 0
+    User = defaultdict(list)
+    Items = set()
+    user_train = {}
+    user_valid = {}
+    user_test = {}
+    # assume user/item index starting from 1
+    f = open("data/%s.txt" % fname, "r")
+
+    for line in f:
+        u, i, t = line.rstrip().split(sep)
+        User[u].append((i, t))
+        Items.add(i)
+
+    for user in User.keys():
+        # sort by time
+        items = sorted(User[user], key=lambda x: x[1])
+        # keep only the items
+        items = [x[0] for x in items]
+        User[user] = items
+        nfeedback = len(User[user])
+        if nfeedback == 1:
+            del User[user]
+            continue
+        elif nfeedback < 3:
+            user_train[user] = User[user]
+            user_valid[user] = []
+            user_test[user] = []
+        else:
+            user_train[user] = User[user][:-2]
+            user_valid[user] = []
+            user_valid[user].append(User[user][-2])
+            user_test[user] = []
+            user_test[user].append(User[user][-1])
+
+    usernum = len(User)
+    itemnum = len(Items)
+    return [user_train, user_valid, user_test, usernum, itemnum]
+
+
+def evaluate(model, dataset, maxlen, num_neg_test):
+    [train, valid, test, usernum, itemnum] = copy.deepcopy(dataset)
```

Comment:
This method expects the parameter `dataset` to be a specific form in order to parse the information by position in the array. For a typical developer, it would be difficult to tell what this `dataset` look like how to prepare such `dataset` (common paint-points in python). I think this is where having a object representation helps:

Imagine you have a `SasrecDataset` object (more details on this in the other comment thread), you can add a type hint (introduced since py3.5) to the parameter to make the function contract more explicit:

```python
def evaluate(model: Sasrec, dataset: SasrecDataset, maxlen: int, num_neg_test: int):
       ....
       usernum = dataset.user_num
       .....
```